keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,681 @@
+
+- `Completed` means that all pods completed, not failed.
+- `Succeeded` means that the job meets the one of successPolicies.
+
+So, the job could have both conditions, `Completed` and `Succeeded`.
+
+#### Surprising whenTargetAchieved behavior if whenTargetAchieved is not specified somewhere  
+
+When the successPolicy has 2 targets and the first target doesn't have `whenTargetAchieved`,
+a `whenTargetAchieved=terminate` is set by default.  
+
+```yaml
+...
+spec:
+  successPolicy:
+    - target:
+        succeededIndexes: "0"
+    - target:
+        succeededIndexes: "1-2"
+        whenTargetAchieved: continue
+```
+
+In that case, if all targets are achieved simultaneously, the first target is evaluated, and all pods are terminated at the first.
+As a result, the second target is not evaluated, and the second `whenTargetAchieved` value is ignored.
+
+This might be surprising behavior for users because the first no specified `whenTargetWhenAchieved` is performed,
+and the second specified `whenTargetWhenAchieved` is not performed.
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.target.succeededIndexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `succeededIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// These policies are evaluated in order, and the job meets one of the successPolicies,
+	// other successPolicies are ignored.
+	// In addition, the order of the successPolicies is immutable.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	// +listType=atomic
+	SuccessPolicy []SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// Each target must have at least one acceptance criterion of
+	// "succeededIndexes" or "count".
+	//
+	// +optional
+	Target *SuccessPolicyTarget
+}
+
+// SuccessPolicyTarget describes a Job can be succeeded based on succeeded indexes.
+type SuccessPolicyTarget struct {
```

Comment:
@alculquicondor and I were just discussing: perhaps it would be useful to add an optional `Reason` field here, which would be applied to the JobCompleted condition when a SuccessPolicy triggers a Job completion. 

This could be useful for high level APIs built on top of the Job API (for example, JobSet https://sigs.k8s.io/jobset) to be able to distinguish between jobs marked completed for different reasons, in order to make more informed, efficient decisions about how to manage those jobs.

In addition, I am submitting a KEP soon for adding an optional `Reason` field to `PodFailurePolicyRule` (https://github.com/kubernetes/enhancements/issues/4443) in order to support certain JobSet failure policy use cases, so it may be useful to support this optional `Reason` field for both success and failure policies.

Reply:
It makes sense. The reason would be helpful.
Here is a new API design agreed with @deads2k:

```yaml
...
  successPolicy:
    criteria:
    - succeededIndexes: "0"
      whenTargetAchieved: terminate
    - succeededIndexes: "1-9"
      whenTargetAchieved: terminate
...
```

If we add a new field, `reason`, I can imagine the below API:

```yaml
...
  successPolicy:
    criteria:
    - succeededIndexes: "0"
      whenTargetAchieved: terminate
      reason: "foo" // when the job meets this criterion, this reason is applied to `Succeeded` condition.
    - succeededIndexes: "1-9"
      whenTargetAchieved: terminate
      reason: "bar" // when the job meets this criterion, this reason is applied to `Succeeded` condition.
...
```

@danielvegamyhre Does this make sense?

Reply:
@tenzen-y I really like that, because if it trickles back into a status or condition, a consuming tool can parse that and give a human readable string back to the running user (and not an exit code or similar).

Reply:
@vsoch I agree with you. Anyway, `reason` field should be reviewed by @deads2k.

Reply:
@danielvegamyhre As I mentioned https://github.com/kubernetes/kubernetes/issues/122972#issuecomment-1924201946, we need to consider alternative approaches, and I don't think that we have enough time for discussions due to the deadline for the enhancement freeze.

So, I think that we should leave the `Reason` field on the beta graduation criterion.
@danielvegamyhre @mimowo @alculquicondor @deads2k WDYT?

Reply:
> Adding it later implies another feature gate. When @deads2k reviews the other KEP, he can probably give guidance of whether it's ok to add the exact same field here.

@alculquicondor @deads2k I see. If we handle the `Reason` field as another featureGate, we might need to track `Reason` feature in another issue.

BTW, we leave `whenCriteriaAchievd` field on the beta graduation. So, we should use another featureGate for `whenCriteriaAchievd`, right?

Reply:
I'm fine with adding a `reason` for clarity.  To clear a up a potential misconception: we can add fields at a beta stage if we discover a need for them, but if we know in advance that we want a field it is important to add that in alpha so that feedback on the API we produce is accurate.

Reply:
It makes sense. I think we should add `reason` field in the alpha stage. So, I will update this KEP.

Reply:
@danielvegamyhre @deads2k @mimowo @alculquicondor I added a `successCriterionMetReasonSuffix` to this proposal.

Reply:
+1, this looks fine to me.

Reply:
I'm closing this thread as the `reason` field is added.